File: javascript/longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
  let longest = 0;
  let start = 0;
  let end = 0;
  const visited = new Set();
  while (end < s.length) {
    const letter = s[end];
    if (visited.has(letter)) {
      while (s[start] !== letter) {
        visited.delete(s[start]);
        start++;
      }
      start++;
    }
    longest = Math.max(longest, end - start + 1);
    visited.add(letter);
    end++;
  }
  return longest;
};

/*
s = "abcabcbb"
          s
           e
letter = b
longest = 3
visited = {a,b,c}
*/

File: python/longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        longest = 0
        currentSubStr = ''
        for letter in s:
          while(currentSubStr.find(letter) is not -1):
            currentSubStr = currentSubStr[1:]
          currentSubStr += letter
          if (longest < len(currentSubStr)):
            longest = len(currentSubStr)
        return longest
